zebra-network/src/peer_set/set.rs

```diff
@@ -220,11 +239,35 @@ where
             Poll::Ready(None) => Err("all peer set background tasks have exited".into()),
         };
 
+        self.shut_down_tasks_and_channels();
+
+        exit_error
```

Comment:
Optional idea: I'm worried that this `poll_background_errors` might become hard to read, because a quick glance seems like the function polls and shuts down the tasks and channels. I'd suggest refactoring this so that the normal flow is more evident. Maybe splitting this method in more methods should help:

```rust
fn poll_background_errors(&mut self, cx: &mut Context) -> Result<(), BoxError> {
    self.receive_guards_if_needed();

    match Pin:::new(&mut self.guards).poll_next(cx) {
        Poll::Pending => Ok(()),
        Poll::Ready(maybe_task_result) => self.shutdown_with_result(
            maybe_task_result.unwrap_or_else(|| {
                Err("all peer set background tasks have exited".into())
            }),
        ),
    }
}
```

Reply:
Fixed in e15e1b152, along with checking the channel during shutdown.